drivers/timer/stm32_lptim_timer.c

```diff
@@ -0,0 +1,306 @@
+	LL_RCC_LSI1_Enable();
+	while (!LL_RCC_LSI1_IsReady()) {
+#else
+	LL_RCC_LSI_Enable();
+	while (!LL_RCC_LSI_IsReady()) {
+#endif /* CONFIG_SOC_SERIES_STM32WBX */
+		/* Wait for LSI ready */
+	}
+
+	LL_RCC_SetLPTIMClockSource(LL_RCC_LPTIM1_CLKSOURCE_LSI);
+
+#else /* CONFIG_STM32_LPTIM_CLOCK_LSI */
+#if defined(LL_APB1_GRP1_PERIPH_PWR)
+	/* Enable the power interface clock */
+	LL_APB1_GRP1_EnableClock(LL_APB1_GRP1_PERIPH_PWR);
+#endif /* LL_APB1_GRP1_PERIPH_PWR */
+	/* enable backup domain */
+	LL_PWR_EnableBkUpAccess();
+	LL_RCC_ForceBackupDomainReset();
+	LL_RCC_ReleaseBackupDomainReset();
+
+	/* enable LSE clock */
+	LL_RCC_LSE_DisableBypass();
+	LL_RCC_LSE_Enable();
+	while (!LL_RCC_LSE_IsReady()) {
+		/* Wait for LSE ready */
+	}
+	LL_RCC_SetLPTIMClockSource(LL_RCC_LPTIM1_CLKSOURCE_LSE);
+
+#endif /* CONFIG_STM32_LPTIM_CLOCK_LSI */
+
+	/* Clear the event flag and possible pending interrupt */
+	IRQ_CONNECT(DT_LPTIM_1_IRQ, DT_LPTIM_1_IRQ_PRI,
+			lptim_irq_handler, 0, 0);
+	irq_enable(DT_LPTIM_1_IRQ);
+
+	/* configure the LPTIM1 counter */
+	LL_LPTIM_SetClockSource(LPTIM1, LL_LPTIM_CLK_SOURCE_INTERNAL);
+	/* configure the LPTIM1 prescaler with 1 */
+	LL_LPTIM_SetPrescaler(LPTIM1, LL_LPTIM_PRESCALER_DIV1);
+	LL_LPTIM_SetPolarity(LPTIM1, LL_LPTIM_OUTPUT_POLARITY_REGULAR);
+	LL_LPTIM_SetUpdateMode(LPTIM1, LL_LPTIM_UPDATE_MODE_IMMEDIATE);
+	LL_LPTIM_SetCounterMode(LPTIM1, LL_LPTIM_COUNTER_MODE_INTERNAL);
+	LL_LPTIM_DisableTimeout(LPTIM1);
+	/* counting start is initiated by software */
+	LL_LPTIM_TrigSw(LPTIM1);
+
+	/* LPTIM1 interrupt set-up before enabling */
+	/* no Compare match Interrupt */
+	LL_LPTIM_DisableIT_CMPM(LPTIM1);
+	LL_LPTIM_ClearFLAG_CMPM(LPTIM1);
+
+	/* Autoreload match Interrupt */
+	LL_LPTIM_EnableIT_ARRM(LPTIM1);
+	LL_LPTIM_ClearFLAG_ARRM(LPTIM1);
+	/* ARROK bit validates the write operation to ARR register */
+	LL_LPTIM_ClearFlag_ARROK(LPTIM1);
+
+	accumulated_lptim_cnt = 0;
+
+	/* Enable the LPTIM1 counter */
+	LL_LPTIM_Enable(LPTIM1);
+
+	/* Set the Autoreload value once the timer is enabled */
+	if (IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		/* LPTIM1 is triggered on a LPTIM_TIMEBASE period */
+		LL_LPTIM_SetAutoReload(LPTIM1, LPTIM_TIMEBASE);
+
+	} else {
+		/* LPTIM1 is triggered on a Tick period */
+		LL_LPTIM_SetAutoReload(LPTIM1, COUNT_PER_TICK);
+	}
+
+	/* Start the LPTIM counter in continuous mode */
+	LL_LPTIM_StartCounter(LPTIM1, LL_LPTIM_OPERATING_MODE_CONTINUOUS);
+
+#ifdef CONFIG_DEBUG
+	/* stop LPTIM1 during DEBUG */
+	LL_DBGMCU_APB1_GRP1_FreezePeriph(LL_DBGMCU_APB1_GRP1_LPTIM1_STOP);
+#endif
+	return 0;
+}
+
+void z_clock_set_timeout(s32_t ticks, bool idle)
+{
+	/* new LPTIM1 AutoReload value to set (aligned on Kernel ticks) */
+	u32_t next_arr = 0;
+
+	ARG_UNUSED(idle);
+
+	/* ARROK bit validates previous write operation to ARR register */
+	while (LL_LPTIM_IsActiveFlag_ARROK(LPTIM1) == 0) {
+	}
+	LL_LPTIM_ClearFlag_ARROK(LPTIM1);
+
+#ifdef CONFIG_TICKLESS_KERNEL
+	if (ticks == K_FOREVER) {
```

Comment:
> So is clear that Kernel will call again the init ?
Isn't it possible that system is waken-up by an external trigger and therefore the kernel just makes a new call to z_clock_set_timeout ?

Can you respond question from @LMESTM ?

Reply:
as specified by the system_timer.h for the Set system clock timeout
K_FOREVER is disabling the timer interrupts, so that in our case we reset the LPTIM timer
>  * Note that ticks can also be passed the special value K_FOREVER,
>  * indicating that no future timer interrupts are expected or required
>  * and that the system is permitted to enter an indefinite sleep even
>  * if this could cause rollover of the internal counter